File: libs/multi-agents-orchestration/tools/sandbox_service.py
#!/usr/bin/env python3
"""
Sandbox Service - Secure code execution service
Watches for sandbox requests and executes them in isolated Docker containers
"""

import json
import time
import subprocess
import os
from pathlib import Path
from datetime import datetime
from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler

# Configuration
REQUESTS_DIR = Path("/workspace/shared/sandbox-requests")
RESULTS_DIR = Path("/workspace/shared/sandbox-results")
PROCESSING_DIR = Path("/workspace/shared/sandbox-processing")
LOG_FILE = Path("/workspace/shared/watcher-logs/sandbox-service.log")

# Default limits
DEFAULT_TIMEOUT = 300  # 5 minutes
DEFAULT_MEMORY = "512m"
DEFAULT_CPU = "1.0"

# Allowed base images (whitelist)
ALLOWED_IMAGES = {
    "python:3.8", "python:3.9", "python:3.10", "python:3.11", "python:3.12",
    "node:14", "node:16", "node:18", "node:20",
    "ruby:3.0", "ruby:3.1", "ruby:3.2",
    "golang:1.19", "golang:1.20", "golang:1.21",
    "openjdk:11", "openjdk:17", "openjdk:21",
    "alpine:latest", "ubuntu:20.04", "ubuntu:22.04"
}

def log(message):
    """Log message with timestamp"""
    timestamp = datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S")
    log_msg = f"[{timestamp}] {message}"
    print(log_msg)
    LOG_FILE.parent.mkdir(parents=True, exist_ok=True)
    with open(LOG_FILE, "a") as f:
        f.write(log_msg + "\n")

def validate_request(request):
    """Validate sandbox request"""
    required_fields = ["request_id", "image", "command"]

    for field in required_fields:
        if field not in request:
            return False, f"Missing required field: {field}"

    if request["image"] not in ALLOWED_IMAGES:
        return False, f"Image not allowed: {request['image']}. Allowed: {', '.join(ALLOWED_IMAGES)}"

    return True, None

def execute_sandbox(request):
    """Execute code in sandbox container"""
    request_id = request["request_id"]
    image = request["image"]
    command = request["command"]
    timeout = request.get("timeout", DEFAULT_TIMEOUT)
    memory = request.get("memory_limit", DEFAULT_MEMORY)
    cpu = request.get("cpu_limit", DEFAULT_CPU)
    working_dir = request.get("working_dir", "/workspace")
    files = request.get("files", {})

    log(f"Executing sandbox request {request_id}")
    log(f"  Image: {image}")
    log(f"  Command: {command}")
    log(f"  Timeout: {timeout}s")

    # Create temporary directory for files
    temp_dir = PROCESSING_DIR / request_id
    temp_dir.mkdir(parents=True, exist_ok=True)

    try:
        # Write files to temp directory
        for filename, content in files.items():
            file_path = temp_dir / filename
            file_path.parent.mkdir(parents=True, exist_ok=True)
            file_path.write_text(content)
            log(f"  Created file: {filename}")

        # Build docker command
        docker_cmd = [
            "docker", "run",
            "--rm",
            "--network", "none",  # No network access
            "--memory", memory,
            "--cpus", cpu,
            "-v", f"{temp_dir}:{working_dir}",
            "-w", working_dir,
            image,
            "sh", "-c", command
        ]

        log(f"  Running: {' '.join(docker_cmd)}")

        # Execute with timeout
        start_time = time.time()
        result = subprocess.run(
            docker_cmd,
            capture_output=True,
            text=True,
            timeout=timeout
        )
        execution_time = time.time() - start_time

        log(f"  Completed in {execution_time:.2f}s")
        log(f"  Exit code: {result.returncode}")

        return {
            "success": result.returncode == 0,
            "exit_code": result.returncode,
            "stdout": result.stdout,
            "stderr": result.stderr,
            "execution_time": execution_time
        }

    except subprocess.TimeoutExpired:
        log(f"  Timeout after {timeout}s")
        return {
            "success": False,
            "error": f"Execution timed out after {timeout} seconds",
            "stdout": "",
            "stderr": "",
            "execution_time": timeout
        }

    except Exception as e:
        log(f"  Error: {str(e)}")
        return {
            "success": False,
            "error": str(e),
            "stdout": "",
            "stderr": "",
            "execution_time": 0
        }

    finally:
        # Cleanup temp directory
        try:
            import shutil
            shutil.rmtree(temp_dir)
            log(f"  Cleaned up temp directory")
        except Exception as e:
            log(f"  Warning: Could not cleanup temp directory: {e}")

def process_request(request_file):
    """Process a sandbox request"""
    try:
        # Read request
        with open(request_file, "r") as f:
            request = json.load(f)

        request_id = request.get("request_id", request_file.stem)
        log(f"Processing request: {request_id}")

        # Validate request
        valid, error = validate_request(request)
        if not valid:
            log(f"  Invalid request: {error}")
            result = {
                "request_id": request_id,
                "success": False,
                "error": error,
                "timestamp": datetime.utcnow().isoformat()
            }
        else:
            # Execute sandbox
            exec_result = execute_sandbox(request)
            result = {
                "request_id": request_id,
                "timestamp": datetime.utcnow().isoformat(),
                **exec_result
            }

        # Write result
        RESULTS_DIR.mkdir(parents=True, exist_ok=True)
        result_file = RESULTS_DIR / f"{request_id}.json"
        with open(result_file, "w") as f:
            json.dump(result, f, indent=2)

        log(f"  Result written to {result_file}")

        # Delete request file
        request_file.unlink()
        log(f"  Request file deleted")

    except Exception as e:
        log(f"Error processing request: {e}")
        import traceback
        log(traceback.format_exc())

class SandboxRequestHandler(FileSystemEventHandler):
    """Handle new sandbox requests"""

    def on_created(self, event):
        if event.is_directory:
            return

        if not event.src_path.endswith(".json"):
            return

        # Small delay to ensure file is fully written
        time.sleep(0.1)

        request_file = Path(event.src_path)
        process_request(request_file)

def main():
    """Main service loop"""
    log("=" * 60)
    log("Sandbox Service Starting")
    log("=" * 60)

    # Create directories
    REQUESTS_DIR.mkdir(parents=True, exist_ok=True)
    RESULTS_DIR.mkdir(parents=True, exist_ok=True)
    PROCESSING_DIR.mkdir(parents=True, exist_ok=True)

    log(f"Requests directory: {REQUESTS_DIR}")
    log(f"Results directory: {RESULTS_DIR}")
    log(f"Allowed images: {', '.join(ALLOWED_IMAGES)}")

    # Process any existing requests
    log("Processing existing requests...")
    existing_requests = list(REQUESTS_DIR.glob("*.json"))
    for request_file in existing_requests:
        log(f"Found existing request: {request_file.name}")
        process_request(request_file)

    # Start watching for new requests
    log("Starting file watcher...")
    event_handler = SandboxRequestHandler()
    observer = Observer()
    observer.schedule(event_handler, str(REQUESTS_DIR), recursive=False)
    observer.start()

    log("Sandbox Service Ready - Waiting for requests...")

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        log("Shutting down...")
        observer.stop()

    observer.join()
    log("Sandbox Service Stopped")

if __name__ == "__main__":
    main()

File: libs/multi-agents-orchestration/tools/sandbox/sandbox_service.js
#!/usr/bin/env node

/**
 * Sandbox Service - Secure code execution service
 * Watches for sandbox requests and executes them in isolated Docker containers
 */

import { promises as fs } from 'fs';
import { spawn } from 'child_process';
import path from 'path';
import { fileURLToPath } from 'url';
import chokidar from 'chokidar';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Configuration
const REQUESTS_DIR = path.resolve('/workspace/shared/sandbox-requests');
const RESULTS_DIR = path.resolve('/workspace/shared/sandbox-results');
const PROCESSING_DIR = path.resolve('/workspace/shared/sandbox-processing');
const LOG_FILE = path.resolve('/workspace/shared/watcher-logs/sandbox-service.log');

// Default limits
const DEFAULT_TIMEOUT = 300; // 5 minutes (in seconds)
const DEFAULT_MEMORY = '512m';
const DEFAULT_CPU = '1.0';

// Allowed base images (whitelist)
const ALLOWED_IMAGES = new Set([
  'python:3.8', 'python:3.9', 'python:3.10', 'python:3.11', 'python:3.12',
  'node:14', 'node:16', 'node:18', 'node:20',
  'ruby:3.0', 'ruby:3.1', 'ruby:3.2',
  'golang:1.19', 'golang:1.20', 'golang:1.21',
  'openjdk:11', 'openjdk:17', 'openjdk:21',
  'alpine:latest', 'ubuntu:20.04', 'ubuntu:22.04'
]);

/**
 * Log message with timestamp
 * @param {string} message - Message to log
 */
async function log(message) {
  const timestamp = new Date().toISOString().replace('T', ' ').substring(0, 19);
  const logMsg = `[${timestamp}] ${message}`;
  console.log(logMsg);

  try {
    await fs.mkdir(path.dirname(LOG_FILE), { recursive: true });
    await fs.appendFile(LOG_FILE, logMsg + '\n', 'utf8');
  } catch (error) {
    console.error(`Failed to write to log file: ${error.message}`);
  }
}

/**
 * Validate sandbox request
 * @param {object} request - Request object
 * @returns {{valid: boolean, error: string|null}}
 */
function validateRequest(request) {
  const requiredFields = ['request_id', 'image', 'command'];

  for (const field of requiredFields) {
    if (!(field in request)) {
      return { valid: false, error: `Missing required field: ${field}` };
    }
  }

  if (!ALLOWED_IMAGES.has(request.image)) {
    return {
      valid: false,
      error: `Image not allowed: ${request.image}. Allowed: ${Array.from(ALLOWED_IMAGES).join(', ')}`
    };
  }

  return { valid: true, error: null };
}

/**
 * Execute command with timeout using child_process
 * @param {string[]} command - Command array
 * @param {number} timeout - Timeout in milliseconds
 * @returns {Promise<{stdout: string, stderr: string, exitCode: number}>}
 */
function executeWithTimeout(command, timeout) {
  return new Promise((resolve, reject) => {
    const [cmd, ...args] = command;
    const process = spawn(cmd, args);

    let stdout = '';
    let stderr = '';
    let timedOut = false;

    const timeoutId = setTimeout(() => {
      timedOut = true;
      process.kill('SIGTERM');

      // Force kill after 5 seconds if not terminated
      setTimeout(() => {
        if (!process.killed) {
          process.kill('SIGKILL');
        }
      }, 5000);
    }, timeout);

    process.stdout.on('data', (data) => {
      stdout += data.toString();
    });

    process.stderr.on('data', (data) => {
      stderr += data.toString();
    });

    process.on('error', (error) => {
      clearTimeout(timeoutId);
      reject(error);
    });

    process.on('close', (exitCode) => {
      clearTimeout(timeoutId);

      if (timedOut) {
        reject(new Error('TIMEOUT'));
      } else {
        resolve({
          stdout,
          stderr,
          exitCode: exitCode || 0
        });
      }
    });
  });
}

/**
 * Recursively remove directory
 * @param {string} dirPath - Directory path
 */
async function removeDirectory(dirPath) {
  try {
    const stat = await fs.stat(dirPath);
    if (!stat.isDirectory()) {
      await fs.unlink(dirPath);
      return;
    }

    const entries = await fs.readdir(dirPath, { withFileTypes: true });

    await Promise.all(
      entries.map(async (entry) => {
        const fullPath = path.join(dirPath, entry.name);
        if (entry.isDirectory()) {
          await removeDirectory(fullPath);
        } else {
          await fs.unlink(fullPath);
        }
      })
    );

    await fs.rmdir(dirPath);
  } catch (error) {
    // Ignore errors if directory doesn't exist
    if (error.code !== 'ENOENT') {
      throw error;
    }
  }
}

/**
 * Execute code in sandbox container
 * @param {object} request - Request object
 * @returns {Promise<object>} Execution result
 */
async function executeSandbox(request) {
  const requestId = request.request_id;
  const image = request.image;
  const command = request.command;
  const timeout = request.timeout || DEFAULT_TIMEOUT;
  const memory = request.memory_limit || DEFAULT_MEMORY;
  const cpu = request.cpu_limit || DEFAULT_CPU;
  const workingDir = request.working_dir || '/workspace';
  const files = request.files || {};

  await log(`Executing sandbox request ${requestId}`);
  await log(`  Image: ${image}`);
  await log(`  Command: ${command}`);
  await log(`  Timeout: ${timeout}s`);

  // Create temporary directory for files
  const tempDir = path.join(PROCESSING_DIR, requestId);
  await fs.mkdir(tempDir, { recursive: true });

  try {
    // Write files to temp directory
    for (const [filename, content] of Object.entries(files)) {
      const filePath = path.join(tempDir, filename);
      await fs.mkdir(path.dirname(filePath), { recursive: true });
      await fs.writeFile(filePath, content, 'utf8');
      await log(`  Created file: ${filename}`);
    }

    // Build docker command
    const dockerCmd = [
      'docker', 'run',
      '--rm',
      '--network', 'none', // No network access
      '--memory', memory,
      '--cpus', cpu,
      '-v', `${tempDir}:${workingDir}`,
      '-w', workingDir,
      image,
      'sh', '-c', command
    ];

    await log(`  Running: ${dockerCmd.join(' ')}`);

    // Execute with timeout
    const startTime = Date.now();

    try {
      const result = await executeWithTimeout(dockerCmd, timeout * 1000);
      const executionTime = (Date.now() - startTime) / 1000;

      await log(`  Completed in ${executionTime.toFixed(2)}s`);
      await log(`  Exit code: ${result.exitCode}`);

      return {
        success: result.exitCode === 0,
        exit_code: result.exitCode,
        stdout: result.stdout,
        stderr: result.stderr,
        execution_time: executionTime
      };

    } catch (error) {
      if (error.message === 'TIMEOUT') {
        await log(`  Timeout after ${timeout}s`);
        return {
          success: false,
          error: `Execution timed out after ${timeout} seconds`,
          stdout: '',
          stderr: '',
          execution_time: timeout
        };
      }
      throw error;
    }

  } catch (error) {
    await log(`  Error: ${error.message}`);
    return {
      success: false,
      error: error.message,
      stdout: '',
      stderr: '',
      execution_time: 0
    };

  } finally {
    // Cleanup temp directory
    try {
      await removeDirectory(tempDir);
      await log('  Cleaned up temp directory');
    } catch (error) {
      await log(`  Warning: Could not cleanup temp directory: ${error.message}`);
    }
  }
}

/**
 * Process a sandbox request
 * @param {string} requestFile - Path to request file
 */
async function processRequest(requestFile) {
  try {
    // Read request
    const content = await fs.readFile(requestFile, 'utf8');
    const request = JSON.parse(content);

    const requestId = request.request_id || path.basename(requestFile, '.json');
    await log(`Processing request: ${requestId}`);

    let result;

    // Validate request
    const validation = validateRequest(request);
    if (!validation.valid) {
      await log(`  Invalid request: ${validation.error}`);
      result = {
        request_id: requestId,
        success: false,
        error: validation.error,
        timestamp: new Date().toISOString()
      };
    } else {
      // Execute sandbox
      const execResult = await executeSandbox(request);
      result = {
        request_id: requestId,
        timestamp: new Date().toISOString(),
        ...execResult
      };
    }

    // Write result
    await fs.mkdir(RESULTS_DIR, { recursive: true });
    const resultFile = path.join(RESULTS_DIR, `${requestId}.json`);
    await fs.writeFile(resultFile, JSON.stringify(result, null, 2), 'utf8');

    await log(`  Result written to ${resultFile}`);

    // Delete request file
    await fs.unlink(requestFile);
    await log('  Request file deleted');

  } catch (error) {
    await log(`Error processing request: ${error.message}`);
    await log(error.stack);
  }
}

/**
 * Main service loop
 */
async function main() {
  await log('='.repeat(60));
  await log('Sandbox Service Starting');
  await log('='.repeat(60));

  // Create directories
  await fs.mkdir(REQUESTS_DIR, { recursive: true });
  await fs.mkdir(RESULTS_DIR, { recursive: true });
  await fs.mkdir(PROCESSING_DIR, { recursive: true });

  await log(`Requests directory: ${REQUESTS_DIR}`);
  await log(`Results directory: ${RESULTS_DIR}`);
  await log(`Allowed images: ${Array.from(ALLOWED_IMAGES).join(', ')}`);

  // Process any existing requests
  await log('Processing existing requests...');
  try {
    const files = await fs.readdir(REQUESTS_DIR);
    const jsonFiles = files.filter(f => f.endsWith('.json'));

    for (const file of jsonFiles) {
      const requestFile = path.join(REQUESTS_DIR, file);
      await log(`Found existing request: ${file}`);
      await processRequest(requestFile);
    }
  } catch (error) {
    await log(`Error processing existing requests: ${error.message}`);
  }

  // Start watching for new requests
  await log('Starting file watcher...');

  const watcher = chokidar.watch(REQUESTS_DIR, {
    ignored: /(^|[\/\\])\../, // ignore dotfiles
    persistent: true,
    ignoreInitial: true, // don't trigger on existing files
    awaitWriteFinish: {
      stabilityThreshold: 100,
      pollInterval: 50
    }
  });

  watcher.on('add', async (filePath) => {
    if (!filePath.endsWith('.json')) {
      return;
    }

    await processRequest(filePath);
  });

  watcher.on('error', async (error) => {
    await log(`Watcher error: ${error.message}`);
  });

  await log('Sandbox Service Ready - Waiting for requests...');

  // Handle graceful shutdown
  process.on('SIGINT', async () => {
    await log('Shutting down...');
    await watcher.close();
    await log('Sandbox Service Stopped');
    process.exit(0);
  });

  process.on('SIGTERM', async () => {
    await log('Shutting down...');
    await watcher.close();
    await log('Sandbox Service Stopped');
    process.exit(0);
  });
}

// Run the service
main().catch(async (error) => {
  await log(`Fatal error: ${error.message}`);
  await log(error.stack);
  process.exit(1);
});
